### ISSUE TITLE
Page and limits

### ISSUE THREAD

**Issue by AndreLobato** [metocean/api-conventions#7]
Do we want to be able to page the queries? 

We probably need to also agree on a request limiter and defaults.

&limit=1000 
or 
&count=1000
?

What should be the expected default?

**Comment by alpha-beta-soup** [comment]
I like `limit`, because of the familiarity in SQL, and `count` is different from a limit in that `count` suggests a measurement of how many things are returned, but `limit` is just a constraint, not a measurement. (e.g. `limit=5000` but only 100 features exist (`count` would be 100, and wouldn't be a query parameter).

I like the idea of paging, but no opinion on how we want to standardise for it.

**Comment by AndreLobato** [comment]
At this point I went for the limit&offset option as...

http://www.django-rest-framework.org/api-guide/pagination/#limitoffsetpagination

**Comment by aportagain** [comment]
I ran into this issue again now with the stats post-processing development. Andre and I discussed this a bit on slack, and I also talked separately to Troy and Dave.

To be honest, I don't see a "clean" way to return cf-json (that's all I'm currently worried about, but of course this discussion could include other data structures / mime-types) in pages...
If I query the MDS and filter by a certain mime-type, I expect the endpoint behind the url template to always return that mime-type, and not for example sometimes cf-json, and sometimes a json object with the first page of data and information on how to find the next one.
Also, how exactly to concatenate / reassemble the individual pages (along which dimension?) isn't necessarily obvious for cf-json in general (it's not tabular / record-based or restricted to 1D data).

Sooo... how about if instead of paging in these situations, we just redirect to an async source? Isn't that process reasonably standardised and therefore likely to work out of the box with many client libraries?

**Comment by aportagain** [comment]
The get_paged_cfjson() method of ops_mds.data_client.MDSDataClient (https://github.com/metocean/ops-mds/blob/master/ops_mds/data_client.py#L80-L118) deals with this for the specific case of pages split along the time dimension, and with fairly simple data structure, and without doing any checks for consistency of that data structure between pages. So pretty fragile, but better than nothing, and seems to be working fine for page responses from the observation API at the moment.

**Comment by hrapizo** [comment]
just as user of the obsapi, I ran into the page limit issue when doing a long query. Solution using opsapi query straight into pandas dataframe was increasing offset and concatenating data, as for example:
def get_data(obsapi, page_size=10000)
```
   offset = 0
   df = DataFrame()
   while True:
       url = obsapi+'&offset='+offset+'&limit='+page_size
       respose = pd.read.csv(url,...).
       df.concatenate(  ... .)
       offset += page_size
       if df.size < page_size:
           break
   return df
```

Alex suggested also using ops-mds data_client:
```
from ops_mds import data_client
m = data_client.MDSDataClient(mds_swagger_url='http://metadata-server-v1-ops-rw.service.consul/v1/swagger.json')
import datetime as dt
predictands = m.get_data_from_dataset('obs_prelud', 'hmo', dt.datetime(2017,12,1), dt.datetime(2018,12,1), 'acenet')
df = predictands.to_dataframe()
```
with the limitation of getting one 'capability' at a time. This would be nice if implemented and could be use for different apis

**Comment by jorgeperezg** [comment]
I've done a small implementation of the loop in the previous comment because I needed it for downloading sofs data.

def get_data(obsapi, page_size=10000):
    df = pd.DataFrame()
    offset = 0
    while offset <= df.size:
        url = '{}&limit={}&offset={}'.format(obsapi, page_size, offset)
        dfi = pd.read_csv(url)
        df = df.append(dfi, ignore_index=True)
        offset += page_size
    return df

obsapi = "https://observations.metoceanapi.com/api/v1/records/?station=sofs&format=csv"
df = get_data(obsapi)
    
    obsapi = "https://observations.metoceanapi.com/api/v1/records/?station=sofs&format=csv"
    df = get_data(obsapi)

**Comment by aportagain** [comment]
Using link headers for pagination seems to have been gaining a bit of traction recently (especially since github started using them). Maybe that's an option we should consider now / for the future?

Here's a few links / examples:
RFC: https://tools.ietf.org/html/rfc5988#section-5
IANA list of link relations (e.g., "prev" and "next"): https://www.iana.org/assignments/link-relations/link-relations.xhtml
github API docs: https://developer.github.com/v3/#link-header
server-side django: https://github.com/tbeadle/django-rest-framework-link-header-pagination
client-side node: https://www.npmjs.com/package/header-pagination
client-side Python requests (doesn't do much...): https://requests.readthedocs.io/en/master/user/advanced/#link-headers

**Comment by AndreLobato** [comment]
Yes Jan suggested same thing, I like that as well at least give us a chance to implement that without breaking the OpenAPI specs

More docs:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Link

**Comment by JBoudMS** [comment]
I'm now discovering this again after @aportagain mentioned this when we came across it in mov-integration-mds so I thought I'd add my 2𝇍.

I prefer the Link based headers approach to ensure we are sticking to the mime-type of a query. In a REST API, I don't expect requesting a resource to return me anything other than the resource.
---
There are some pretty good docs about how to do this in a [GraphQL API](https://graphql.org/learn/pagination/). It's not like-for-like given the differences with REST, but we might still get some useful details out of it.

They return the pagination details as part of their response, and they recommend using cursor based pagination and making the token opaque. e.g., `first={number}&after=${token}`. This is because it gives you additional flexibility to change the pagination in the future. This would also allow you to use timestamps or other ids to ensure alignment of the pages.

We could adapt this to REST by using the same functionality but returning the Pagination inside the Link header.
---
The Global Lightning API uses `limit` and `offset` and returns a `prev` and `next` in the Link header to return the relevant details. It works really well for us for historic/settled data (older than 5 minutes). It has some issues with recent data as the order of the strikes can change. More specifically it has two issues:
1. It can return the same strike twice (if there are new strikes in previous pages)
2. It can miss strikes (if there are new strikes in previous pages)

Cursor based pagination would avoid the first issue since it can ensure the start of the page is the end of the previous page, but can't resolve the second.

With regards to CF-JSON, I'd personally only ever expect it to be split down the time dimension. That might not be the most efficient split if we have a large grid with a lot of variables and a short time query for instance but it's what makes the most intuitive sense for pagination of such a request to me. I doubt it is an issue since I doubt the CF-JSON is being updated this often, but cursor based pagination would allow you to make sure that the pages are aligned between requests.
